grpc/disruption_listener_service.go

```diff
@@ -0,0 +1,126 @@
+		"UNAUTHENTICATED":     codes.Unauthenticated,
+	}
+)
+
+// SendDisruption makes a gRPC call to DisruptionListener
+func (d *DisruptionListener) SendDisruption(ctx context.Context, ds *pb.DisruptionSpec) (*emptypb.Empty, error) {
+	if ds == nil {
+		log.Fatal("Cannot execute SendDisruption when DisruptionSpec is nil")
+	}
+
+	for _, alt := range ds.EndpointAlterations {
+		if alt.TargetEndpoint == "" {
+			log.Print("SendDisruption does not specify TargetEndpoint for all endpointAlterations")
+			return nil, status.Error(codes.InvalidArgument, "Cannot execute SendDisruption without specifying TargetEndpoint for all endpointAlterations")
+		}
+		if alt.ErrorToReturn == "" && alt.OverrideToReturn == "" {
+			log.Print("For at least one endpointAlteration, SendDisruption specifies neither ErrorToReturn nor OverrideToReturn")
+			return nil, status.Error(codes.InvalidArgument, "Cannot execute SendDisruption without specifying either ErrorToReturn or OverrideToReturn for all target endpoints")
+		}
+
+		alteration := EndpointAlteration{
+			TargetEndpoint:   alt.TargetEndpoint,
+			ErrorToReturn:    alt.ErrorToReturn,
+			OverrideToReturn: alt.OverrideToReturn,
+		}
+		d.Spec = append(d.Spec, alteration)
+	}
+
+	return &emptypb.Empty{}, nil
+}
+
+// DisruptionStatus lists all configured endpoint alterations for DisruptionListener
+func (d *DisruptionListener) DisruptionStatus(context.Context, *emptypb.Empty) (*pb.DisruptionSpec, error) {
+	ds := pb.DisruptionSpec{EndpointAlterations: []*pb.EndpointAlteration{}}
+
+	for _, alt := range d.Spec {
+		ds.EndpointAlterations = append(
+			ds.EndpointAlterations,
+			&pb.EndpointAlteration{
+				TargetEndpoint:   alt.TargetEndpoint,
+				ErrorToReturn:    alt.ErrorToReturn,
+				OverrideToReturn: alt.OverrideToReturn,
+			},
+		)
+	}
+	return &ds, nil
+}
+
+// CleanDisruption removes all configured endpoint alterations for DisruptionListener
+func (d *DisruptionListener) CleanDisruption(context.Context, *emptypb.Empty) (*emptypb.Empty, error) {
+	d.Spec = []EndpointAlteration{}
+	return &emptypb.Empty{}, nil
+}
+
+// ChaosServerInterceptor is a function which can be registered on instantiation of a gRPC server
+// to intercept all traffic to the server and crosscheck their endpoints to disrupt them
+func (d *DisruptionListener) ChaosServerInterceptor(ctx context.Context, req interface{},
+	info *grpc.UnaryServerInfo, handler grpc.UnaryHandler) (response interface{}, err error) {
+	// Find out what api endpoint the request is for
+	// FullMethod is the full RPC method string, i.e., /package.service/method.
+	endpoint := info.FullMethod
+
+	for _, alteration := range d.Spec {
+		if endpoint == alteration.TargetEndpoint {
+			if alteration.ErrorToReturn != "" {
+				log.Printf("Error Code: %s", errorMap[alteration.ErrorToReturn])
+				return nil, status.Error(errorMap[alteration.ErrorToReturn], fmt.Sprintf("Chaos-Controller injected this error: %s", alteration.ErrorToReturn))
```

Comment:
in the future we should check with users if they want this error message [not really sure if these errors normally just get blank messages or what, just a thought, not a big concern]